### PR TITLE
CMS: Adding signers should gen a content type attribute

### LINKED ISSUE #8923: openssl cms -resign dosn't work properly
When I use the cms command with the switch -resign to add another signature to a signed file, the content-type attribute is non present on the added signature. This is an error because the rfc 5652 par 11.1 (Content Type) require that "The content-type attribute type MUST be present whenever signed attributes are present in signed-data".
The problem is always reproducible and is detected, for example, from Bouncycastle.
The commands I've executed are:

$ openssl cms -sign -binary -nodetach -stream -nosmimecap -md SHA256 -signer keysig_1.pem -passin pass:12345678 -in hello.txt -outform DER -out hello111.txt.p7m
$ openssl cms -resign -binary -nodetach -nosmimecap -md SHA256 -signer keysig_2.pem -passin pass:12345678 -inform DER -in hello111.txt.p7m -outform DER -out hello111.txt.p7m.p7m

### PR DESCRIPTION
**Pull request by slontis**
Fixes #8923

Found using the openssl cms -resign option.
This uses an alternate path to do the signing which was not adding the required signed attribute
content type. This attribute should always exist since it is required is there are any signed attributes.
As the signing time attribute is always added in code, then the content type attribute is also always required.

##### Checklist

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by slontis** [comment]
#8117 needs to be merged for verify -cades to work after -resigning.

File: test/recipes/80-test_cms.t [hunk 1]
```diff
@@ -509,6 +509,50 @@ subtest "CMS <=> CMS consistency tests, modified key parameters\n" => sub {
...
+# Returns the number of matches of a Content Type Attribute in a binary file.
+sub contentType_matches {
+  # Read in a binary file
+  my ($in) = @_;
+  open (HEX_IN, "$in") or die("open failed for $in : $!");
+  binmode(HEX_IN);
+  local $/;
+  my $str = <HEX_IN>;
+
+  #Find ASN1 data for a Content Type Attribute (with a OID od PKCS7 data)
```
Comment (FdaSilvaYY) [inline test/recipes/80-test_cms.t]: "with a. OID __od__ PKCS7 data" ?

Comment (slontis) [reply test/recipes/80-test_cms.t]: 'd' is right next to 'f' on the keyboard.

**Comment by slontis** [comment]
@ping

**Comment by ping** [comment]
Hi @slontis, I've been getting frequent mentions from you on this repo over the last few days. Please know that the "ping" handle belongs to a real GitHub user who will get notifications every time you do an @/ping. Thanks.

**Comment by slontis** [comment]
@ping Sorry about that - will try not to spam you in future.
I need to not put the @ in front of the keyword :)

File: crypto/cms/cms_att.c [hunk 2]
```diff
@@ -149,4 +199,86 @@ void *CMS_unsigned_get0_data_by_OBJ(CMS_SignerInfo *si, ASN1_OBJECT *oid,
     return X509at_get0_data_by_OBJ(si->unsignedAttrs, oid, lastpos, type);
 }
 
-/* Specific attribute cases */
+/*
+ * Retrieve an attribute by nid from a stack of attributes starting at index
+ * *lastpos.
+ * Returns the attribute or NULL if there is no attribute.
+ * *lastpos returns the index of the found attribute or -1 if not found.
```
Comment (mattcaswell) [inline crypto/cms/cms_att.c]: Where is *lastpos set to -1 if not found?

File: crypto/cms/cms_att.c [hunk 3]
```diff
@@ -149,4 +199,86 @@ void *CMS_unsigned_get0_data_by_OBJ(CMS_SignerInfo *si, ASN1_OBJECT *oid,
...
+    at = X509at_get_attr(attrs, loc);
+    *lastpos = loc;
+    return at;
+}
+
+static int cms_check_attribute(int nid, int flags, int type,
+                               const STACK_OF(X509_ATTRIBUTE) *attrs,
+                               int have_attrs)
+{
+    int lastpos = -1;
```
Comment (mattcaswell) [inline crypto/cms/cms_att.c]: Shouldn't this really start at index 0?

File: crypto/cms/cms_att.c [hunk 4]
```diff
@@ -149,4 +199,86 @@ void *CMS_unsigned_get0_data_by_OBJ(CMS_SignerInfo *si, ASN1_OBJECT *oid,
...
+    X509_ATTRIBUTE *at = cms_attrib_get(nid, attrs, &lastpos);
+
+    if (at != NULL) {
+        int count = X509_ATTRIBUTE_count(at);
+
+        /* Is this attribute allowed? */
+        if (!(flags & type)
+            /* check if multiple attributes of the same type are allowed */
+            || ((flags & CMS_ATTR_F_ONLY_ONE)
```
Comment (mattcaswell) [inline crypto/cms/cms_att.c]: Preferred style is: `(flags & CMS_ATTR_F_ONLY_ONE) != 0`, Similarly with other boolean `&` expressions in this function.

File: crypto/cms/cms_att.c (hunk 3 above)
Comment (slontis) [reply crypto/cms/cms_att.c]: The X509 method that it calls internally does an autoincrement of the pos (so that it starts at 0). Not my doing :)

**Comment by slontis** [comment]
Last commit contains the changes + updated commit messages

**Comment by slontis** [comment]
Merged to master and 1_1_1

**Pull request closed without merging**